File: src/main/java/at/hannibal2/skyhanni/config/features/combat/endernode/EnderNodeConfig.java
package at.hannibal2.skyhanni.config.features.combat.endernode;

import at.hannibal2.skyhanni.config.FeatureToggle;
import at.hannibal2.skyhanni.config.core.config.Position;
import com.google.gson.annotations.Expose;
import io.github.moulberry.moulconfig.annotations.ConfigEditorBoolean;
import io.github.moulberry.moulconfig.annotations.ConfigEditorDraggableList;
import io.github.moulberry.moulconfig.annotations.ConfigOption;
import io.github.moulberry.moulconfig.observer.Property;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static at.hannibal2.skyhanni.config.features.combat.endernode.EnderNodeDisplayEntry.COINS_MADE;
import static at.hannibal2.skyhanni.config.features.combat.endernode.EnderNodeDisplayEntry.ENCHANTED_ENDER_PEARL;
import static at.hannibal2.skyhanni.config.features.combat.endernode.EnderNodeDisplayEntry.ENCHANTED_END_STONE;
import static at.hannibal2.skyhanni.config.features.combat.endernode.EnderNodeDisplayEntry.ENCHANTED_OBSIDIAN;
import static at.hannibal2.skyhanni.config.features.combat.endernode.EnderNodeDisplayEntry.ENDERMAN_PET;
import static at.hannibal2.skyhanni.config.features.combat.endernode.EnderNodeDisplayEntry.ENDERMITE_NEST;
import static at.hannibal2.skyhanni.config.features.combat.endernode.EnderNodeDisplayEntry.ENDER_ARMOR;
import static at.hannibal2.skyhanni.config.features.combat.endernode.EnderNodeDisplayEntry.GRAND_XP_BOTTLE;
import static at.hannibal2.skyhanni.config.features.combat.endernode.EnderNodeDisplayEntry.MAGICAL_RUNE_I;
import static at.hannibal2.skyhanni.config.features.combat.endernode.EnderNodeDisplayEntry.MITE_GEL;
import static at.hannibal2.skyhanni.config.features.combat.endernode.EnderNodeDisplayEntry.NODES_MINED;
import static at.hannibal2.skyhanni.config.features.combat.endernode.EnderNodeDisplayEntry.SHRIMP_THE_FISH;
import static at.hannibal2.skyhanni.config.features.combat.endernode.EnderNodeDisplayEntry.SPACER_1;
import static at.hannibal2.skyhanni.config.features.combat.endernode.EnderNodeDisplayEntry.SPACER_2;
import static at.hannibal2.skyhanni.config.features.combat.endernode.EnderNodeDisplayEntry.TITANIC_XP_BOTTLE;
import static at.hannibal2.skyhanni.config.features.combat.endernode.EnderNodeDisplayEntry.TITLE;

public class EnderNodeConfig {
    @Expose
    @ConfigOption(
        name = "Enabled",
        desc = "Tracks all of your drops from mining Ender Nodes in the End.\n" +
            "Also tracks drops from Endermen."
    )
    @ConfigEditorBoolean
    @FeatureToggle
    public boolean enabled = false;

    @Expose
    @ConfigOption(
        name = "Text Format",
        desc = "Drag text to change the appearance of the overlay."
    )
    @ConfigEditorDraggableList()
    public Property<List<EnderNodeDisplayEntry>> textFormat = Property.of(new ArrayList<>(Arrays.asList(
        TITLE,
        NODES_MINED,
        COINS_MADE,
        SPACER_1,
        ENDERMITE_NEST,
        ENCHANTED_END_STONE,
        ENCHANTED_OBSIDIAN,
        ENCHANTED_ENDER_PEARL,
        GRAND_XP_BOTTLE,
        TITANIC_XP_BOTTLE,
        MAGICAL_RUNE_I,
        MITE_GEL,
        SHRIMP_THE_FISH,
        SPACER_2,
        ENDER_ARMOR,
        ENDERMAN_PET)
    ));

    @Expose
    public Position position = new Position(10, 80, false, true);
}

File: src/main/java/at/hannibal2/skyhanni/config/features/combat/end/EnderNodeConfig.kt
package at.hannibal2.skyhanni.config.features.combat.end

import at.hannibal2.skyhanni.config.FeatureToggle
import at.hannibal2.skyhanni.config.core.config.Position
import at.hannibal2.skyhanni.features.combat.end.endernodetracker.EnderNode
import com.google.gson.annotations.Expose
import io.github.notenoughupdates.moulconfig.annotations.ConfigEditorBoolean
import io.github.notenoughupdates.moulconfig.annotations.ConfigEditorDraggableList
import io.github.notenoughupdates.moulconfig.annotations.ConfigLink
import io.github.notenoughupdates.moulconfig.annotations.ConfigOption
import io.github.notenoughupdates.moulconfig.observer.Property

class EnderNodeConfig {
    @Expose
    @ConfigOption(
        name = "Enabled",
        desc = "Tracks all of your drops from mining Ender Nodes in the End.\n" +
            "Also tracks drops from Endermen."
    )
    @ConfigEditorBoolean
    @FeatureToggle
    var enabled: Boolean = false

    @Expose
    @ConfigOption(
        name = "Only While Holding Tool",
        desc = "Only show the tracker if holding a pickaxe, drill or gauntlet in hand."
    )
    @ConfigEditorBoolean
    var onlyPickaxe: Boolean = false

    @Expose
    @ConfigOption(name = "Text Format", desc = "Drag text to change the appearance of the overlay.")
    @ConfigEditorDraggableList
    val textFormat: Property<MutableList<EnderNodeDisplayEntry>> = Property.of(
        mutableListOf(
            EnderNodeDisplayEntry.TITLE,
            EnderNodeDisplayEntry.NODES_MINED,
            EnderNodeDisplayEntry.COINS_MADE,
            EnderNodeDisplayEntry.SPACER_1,
            EnderNodeDisplayEntry.ENDERMITE_NEST,
            EnderNodeDisplayEntry.ENCHANTED_END_STONE,
            EnderNodeDisplayEntry.ENCHANTED_OBSIDIAN,
            EnderNodeDisplayEntry.ENCHANTED_ENDER_PEARL,
            EnderNodeDisplayEntry.GRAND_XP_BOTTLE,
            EnderNodeDisplayEntry.TITANIC_XP_BOTTLE,
            EnderNodeDisplayEntry.MAGICAL_RUNE_I,
            EnderNodeDisplayEntry.MITE_GEL,
            EnderNodeDisplayEntry.SHRIMP_THE_FISH,
            EnderNodeDisplayEntry.SPACER_2,
            EnderNodeDisplayEntry.ENDER_ARMOR,
            EnderNodeDisplayEntry.ENDERMAN_PET
        )
    )

    enum class EnderNodeDisplayEntry(private val displayName: String) {
        TITLE("§5§lEnder Node Tracker"),
        NODES_MINED("§d1,303 Ender Nodes Mined"),
        COINS_MADE("§615.3M Coins Made"),
        SPACER_1(" "),
        ENDERMITE_NEST("§b123 §cEndermite Nest"),
        ENCHANTED_END_STONE("§b832 §aEnchanted End Stone"),
        ENCHANTED_OBSIDIAN("§b230 §aEnchanted Obsidian"),
        ENCHANTED_ENDER_PEARL("§b1630 §aEnchanted Ender Pearl"),
        GRAND_XP_BOTTLE("§b85 §aGrand Experience Bottle"),
        TITANIC_XP_BOTTLE("§b4 §9Titanic Experience Bottle"),
        END_STONE_SHULKER("§b15 §9End Stone Shulker"),
        END_STONE_GEODE("§b53 §9End Stone Geode"),
        MAGICAL_RUNE_I("§b10 §d◆ Magical Rune I"),
        ENDER_GAUNTLET("§b24 §5Ender Gauntlet"),
        MITE_GEL("§b357 §5Mite Gel"),
        SHRIMP_THE_FISH("§b2 §cShrimp The Fish"),
        SPACER_2(" "),
        ENDER_ARMOR("§b200 §5Ender Armor"),
        ENDER_HELMET("§b24 §5Ender Helmet"),
        ENDER_CHESTPLATE("§b24 §5Ender Chestplate"),
        ENDER_LEGGINGS("§b24 §5Ender Leggings"),
        ENDER_BOOTS("§b24 §5Ender Boots"),
        ENDER_NECKLACE("§b24 §5Ender Necklace"),
        ENDERMAN_PET("§f10§7-§a8§7-§93§7-§52§7-§61 §fEnderman Pet"),
        ;

        override fun toString() = displayName

        companion object {
            private val enderNodeCache: MutableMap<EnderNodeDisplayEntry, EnderNode?> = mutableMapOf()
        }

        fun toEnderNodeOrNull(): EnderNode? = enderNodeCache.getOrPut(this) {
            EnderNode.entries.firstOrNull {
                it.toEnderNodeDisplayEntryOrNull() == this
            }
        }
    }

    @Expose
    @ConfigLink(owner = EnderNodeConfig::class, field = "enabled")
    val position: Position = Position(10, 80)
}
